{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# Ray-finned fish tree from TimeTree - species level (Actinopterygii)\n",
    "\n",
    "The goal of this notebook is to produce a clean species tree that is annotated\n",
    "with orders and families in internal node names. It also produces an order-level\n",
    "tree with the orders as the leaves that will serve as a scaffold in successive\n",
    "notebooks.\n",
    "\n",
    "To accomplish this, we need to look up the order and family for every species in\n",
    "the tree. This requires some cleaning of the tree, which is removing taxa that\n",
    "cannot be looked up and fixing node names so lookups can be done properly. It\n",
    "also involves cleaning of the database, which is an iterative process. The final\n",
    "version of this notebook is after all this has been done. The taxonomy database\n",
    "is saved locally during this process and the final db has been tweaked quite a\n",
    "bit.\n",
    "\n",
    "This script requires ETE, which normally does not install properly on Windows.\n",
    "You can run it in Windows, but you have to install ETE carefully and with the\n",
    "right prerequisites. You may have issues running a blanket `import ete`, as this\n",
    "pulls in all kinds of things that don't work on windows. Here I only use the\n",
    "`Tree` class, which appears to work."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's use my handy dandy taxonomy database to look up order and family info.\n",
    "\n",
    "import TaxonomyDB as taxonomydb\n",
    "from importlib import reload\n",
    "reload(taxonomydb)\n",
    "\n",
    "db = taxonomydb.TaxonomyDB()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2",
   "metadata": {},
   "source": [
    "Quick database test..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "db.get_species_info(\"Calamus\")\n",
    "db.get_species_info(\"Decodon\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4",
   "metadata": {},
   "source": [
    "Database cleanup. Problematic taxa identified on previous runs. This only needs to be run once as the DB caches changes locally. Running it multiple times doesn't hurt anything, you just don't need to do it.\n",
    "\n",
    "You **do** need to run it your first time through, however, otherwise incorrect GBIF lookups will be used. This block below basically pre-loads the local database with fixed taxonomy info that is incorrect in GBIF online. Well, maybe not incorrect, but not useable for fish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extra taxa not in GBIF or NCBI that we need to handle.\n",
    "db.remove_species_info(genus='Calamus')\n",
    "db.remove_species_info(genus='Ophidion')\n",
    "db.remove_species_info(genus='Liparis')\n",
    "db.remove_species_info(genus='Orestias')\n",
    "db.remove_species_info(genus='Scuticaria')\n",
    "db.remove_species_info(genus='Centropogon')\n",
    "db.remove_species_info(genus='Exostoma')\n",
    "db.remove_species_info(genus='Mora')\n",
    "db.remove_species_info(genus='Bullockia')\n",
    "db.remove_species_info(genus='Symphysodon')\n",
    "db.remove_species_info(genus='Scleronema')\n",
    "db.remove_species_info(genus='Histrio')\n",
    "db.remove_species_info(genus='Nansenia')\n",
    "db.remove_species_info(genus='Olyra')\n",
    "db.remove_species_info(genus='Cynodon')\n",
    "db.remove_species_info(genus='Bairdiella')\n",
    "db.remove_species_info(genus='Ranzania')\n",
    "db.remove_species_info(genus='Zeus')\n",
    "db.remove_species_info(genus='Argentina')\n",
    "db.remove_species_info(genus='Dactylophora')\n",
    "db.remove_species_info(genus='Trachystoma')\n",
    "db.remove_species_info(genus='Coris')\n",
    "db.remove_species_info(genus='Decodon')\n",
    "db.remove_species_info(genus='Trichodon')\n",
    "db.remove_species_info(genus='Rondeletia')\n",
    "db.remove_species_info(genus='Mallotus')\n",
    "db.remove_species_info(genus='Rhaphiodon')\n",
    "db.remove_species_info(genus='Heterotis')\n",
    "\n",
    "db.add_species_info('Calamus', 'Acanthuriformes', 'Sparidae')\n",
    "db.add_species_info('Ophidion', 'Ophidiiformes', 'Ophidiidae')\n",
    "db.add_species_info('Liparis', 'Perciformes', 'Liparidae')\n",
    "db.add_species_info('Orestias', 'Cyprinodontiformes', 'Cyprinodontidae')\n",
    "db.add_species_info('Scuticaria', 'Anguilliformes', 'Muraenidae')\n",
    "db.add_species_info('Centropogon', 'Perciformes', 'Syngnathidae')\n",
    "db.add_species_info('Exostoma', 'Siluriformes', 'Sisoridae')\n",
    "db.add_species_info('Mora', 'Gadiformes', 'Moridae')\n",
    "db.add_species_info('Bullockia', 'Siluriformes', 'Tricomycteridae')\n",
    "db.add_species_info('Symphysodon', 'Cichliformes', 'Cichlidae')\n",
    "db.add_species_info('Scleronema', 'Siluriformes', 'Trichomycteridae')\n",
    "db.add_species_info('Histrio', 'Lophiiformes', 'Antennariidae')\n",
    "db.add_species_info('Nansenia', 'Argentiniformes', 'Microstomatidae')\n",
    "db.add_species_info('Olyra', 'Siluriformes', 'Bagridae')\n",
    "db.add_species_info('Cynodon', 'Characiformes', 'Cynodontidae')\n",
    "db.add_species_info('Bairdiella', 'Acanthuriformes', 'Sciaenidae')\n",
    "db.add_species_info('Ranzania', 'Tetraodontiformes', 'Molidae')\n",
    "db.add_species_info('Zeus', 'Zeiformes', 'Zeidae')\n",
    "db.add_species_info('Argentina', 'Argentiniformes', 'Argentinidae')\n",
    "db.add_species_info('Dactylophora', 'Centrarchiformes', 'Cheilodactylidae')\n",
    "db.add_species_info('Trachystoma', 'Mugiliformes', 'Mugilidae')\n",
    "db.add_species_info('Coris', 'Labriformes', 'Labridae')\n",
    "db.add_species_info('Decodon', 'Labriformes', 'Labridae')\n",
    "db.add_species_info('Trichodon', 'Perciformes', 'Trichodontidae')\n",
    "db.add_species_info('Rondeletia', 'Beryciformes', 'Rondeletiidae')\n",
    "db.add_species_info('Mallotus', 'Osmeriformes', 'Osmidae')\n",
    "db.add_species_info('Rhaphiodon', 'Characiformes', 'Cynodontidae')\n",
    "db.add_species_info('Heterotis', 'Osteoglossiformes', 'Arapaimidae')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in the trees.\n",
    "\n",
    "from ete3 import Tree\n",
    "\n",
    "species_tree = Tree('ray-finned fishes_species.nwk', format=1, quoted_node_names=True)\n",
    "\n",
    "# Output the number of tips in each tree.\n",
    "print(f'Species tree has {len(species_tree.get_leaves())} tips.')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7",
   "metadata": {},
   "source": [
    "## Data cleaning\n",
    "\n",
    "Trees from timetree.org are a composite of many trees from different studies. Some of these have odd naming schemes and don't conform to \"Genus species\" naming like we need here for lookups. This section of code cleans up names so that lookups for family and order actually work.\n",
    "\n",
    "Every dataset is a little different, and the code blocks below are specific to *this* dataset. Cleanup of the insect tree was quite different, as there were many misnamed genera and even some bacteria included. Oh, and some fungi, and a couple of bivalves."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# There are a few family names posing as genera. Remove anything that\n",
    "# ends with \"idae\" or \"inae\".\n",
    "\n",
    "import re\n",
    "\n",
    "for leaf in species_tree.get_leaves():\n",
    "    # Grab the first part of the name. It can be the first word separated by a space,\n",
    "    # underscore, or dash.\n",
    "    genus_name = re.split(r'[ _-]', leaf.name)[0]\n",
    "\n",
    "    # If the genus name ends with \"idae\" or \"inae\", remove it.\n",
    "    if genus_name.endswith(\"idae\") or genus_name.endswith(\"inae\"):\n",
    "        print(f\"Removing family name posing as genus: {genus_name}\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].detach()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Are there any leaves that are just a number, or just an underscore and a number?\n",
    "for leaf in species_tree.get_leaves():\n",
    "    if re.match(r'^_?\\d+$', leaf.name):\n",
    "        print(f\"Removing invalid leaf name: {leaf.name}\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].detach()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Misspellings. \n",
    "\n",
    "for leaf in species_tree.get_leaves():\n",
    "    # Labeoarbus -> Labeobarbus\n",
    "    if leaf.name.startswith(\"Labeoarbus\"):\n",
    "        print(f\"Correcting misspelled genus: {leaf.name} -> Labeobarbus\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].name = leaf.name.replace(\"Labeoarbus\", \"Labeobarbus\")\n",
    "\n",
    "    # Tariqlabeo -> Tariqilabeo\n",
    "    if leaf.name.startswith(\"Tariqlabeo\"):\n",
    "        print(f\"Correcting misspelled genus: {leaf.name} -> Tariqilabeo\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].name = leaf.name.replace(\"Tariqlabeo\", \"Tariqilabeo\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Inactive taxa.\n",
    "#Coelonotus_leiaspis should be Microphis_leiaspis\n",
    "for leaf in species_tree.get_leaves():\n",
    "    if leaf.name.startswith(\"Coelonotus_leiaspis\"):\n",
    "        print(f\"Correcting inactive taxon: {leaf.name} -> Microphis_leiaspis\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].name = leaf.name.replace(\"Coelonotus_leiaspis\", \"Microphis_leiaspis\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's make sure there are no duplicated names in the tree. Check all tips - the internal nodes are not important for this.\n",
    "# Print the list of duplicated names, if any. Also print how many times each duplicated name appears.\n",
    "tip_names = [leaf.name for leaf in species_tree.iter_leaves()]\n",
    "duplicated_names = set([name for name in tip_names if tip_names.count(name) > 1])\n",
    "print(\"Duplicated names and their counts:\")\n",
    "for name in duplicated_names:\n",
    "    print(f\"{name}: {tip_names.count(name)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# There's only one, Tenebrosternarchus_preto. Get the list of all leaves with that name.\n",
    "tenebrosternarchus_leaves = species_tree.search_nodes(name=\"Tenebrosternarchus_preto\")\n",
    "print(f\"Found {len(tenebrosternarchus_leaves)} leaves with the name 'Tenebrosternarchus_preto'.\")\n",
    "\n",
    "# Rename them to Tenebrosternarchus_preto_1 and Tenebrosternarchus_preto_2.\n",
    "count = 1\n",
    "for leaf in species_tree.get_leaves():\n",
    "    if leaf.name == \"Tenebrosternarchus_preto\":\n",
    "        new_name = f\"Tenebrosternarchus_preto_{count}\"\n",
    "        print(f\"Renaming duplicate species: {leaf.name} -> {new_name}\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].name = new_name\n",
    "        count += 1\n",
    "\n",
    "# Check for duplicates again.\n",
    "tip_names = [leaf.name for leaf in species_tree.iter_leaves()]\n",
    "duplicated_names = set([name for name in tip_names if tip_names.count(name) > 1])\n",
    "print(\"Duplicated names after renaming:\", duplicated_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A couple of names are quoted because there are spaces in the name. Remove the quotes\n",
    "# and replace spaces with underscores. Also remove any commas, periods, and parenthesis, and colons.\n",
    "for leaf in species_tree.get_leaves():\n",
    "    cleaned_name = leaf.name.replace(\" \", \"_\").replace(\",\", \"\").replace(\".\", \"\").replace(\"(\", \"\").replace(\")\", \"\").replace(\":\", \"\")\n",
    "    if cleaned_name != leaf.name:\n",
    "        print(f\"Cleaning leaf name: {leaf.name} -> {cleaned_name}\")\n",
    "        species_tree.search_nodes(name=leaf.name)[0].name = cleaned_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lookups. Try to look up each genus in the taxonomy database.\n",
    "\n",
    "############\n",
    "##\n",
    "## IMPORTANT NOTE - the first time this is run, it will dump a lot of info and also take about 20-30 minutes.\n",
    "##                  Subsequent runs will be much faster (a couple of seconds) as the db is cached locally.\n",
    "##\n",
    "############\n",
    "\n",
    "import re\n",
    "\n",
    "leaves_sorted_by_order = {}\n",
    "for leaf in species_tree.get_leaves():\n",
    "    # Grab the first part of the name. It can be the first word separated by a space,\n",
    "    # underscore, or dash.\n",
    "    genus_name = re.split(r'[ _-]', leaf.name)[0]\n",
    "\n",
    "    # Get the order name from the genus name.\n",
    "    order_name, family = db.get_species_info(genus_name)\n",
    "\n",
    "    # Add the order and family to each leaf.\n",
    "    leaf.add_feature('order', order_name)\n",
    "    leaf.add_feature('family', family)\n",
    "\n",
    "    # Make a list of all the insect orders in the tree.\n",
    "    if order_name not in leaves_sorted_by_order:\n",
    "        leaves_sorted_by_order[order_name] = []\n",
    "        print(f\"Adding new order: {order_name}\")\n",
    "    leaves_sorted_by_order[order_name].append(leaf)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's make a set of all the orders we found. For each leaf, get the order feature\n",
    "# and add it to the set\n",
    "\n",
    "found_orders = set()\n",
    "for leaf in species_tree.get_leaves():\n",
    "    found_orders.add(leaf.order)\n",
    "print(f\"Found {len(found_orders)} unique orders in the tree: {found_orders}\")\n",
    "for order in sorted(found_orders):\n",
    "    print(f\"- {order}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find everything that does not end in \"formes\". These are likely to be misclassified or missing orders.\n",
    "print(\"Leaves with orders that do not end in 'formes':\")\n",
    "for leaf in species_tree.get_leaves():\n",
    "    if not leaf.order.endswith(\"formes\"):\n",
    "        print(f\"- {leaf.name} (order: {leaf.order})\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For looking up specific orders, e.g., Brassicales, that are not fish orders. These\n",
    "# are fixed above in the database cleanup section.\n",
    "for leaf in species_tree.get_leaves():\n",
    "    if leaf.order == \"Myrtales\":\n",
    "        print(f\"Leaf in order Myrtales: {leaf.name}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19",
   "metadata": {},
   "source": [
    "## Monophyletic groups\n",
    "\n",
    "Every leaf has an order and family added. Let's find the order-level nodes in the tree and name those."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each insect order, find the corresponding leaves in the genus tree.\n",
    "for order_name in found_orders:\n",
    "    leaves_in_given_order = species_tree.search_nodes(order=order_name)\n",
    "    print(f\"Order {order_name} has {len(leaves_in_given_order)} leaves in the species tree.\")\n",
    "\n",
    "    # Now get the most recent common ancestor for these leaves.\n",
    "    if leaves_in_given_order:\n",
    "        mrca = species_tree.get_common_ancestor(leaves_in_given_order)\n",
    "        print(f\"The most recent common ancestor of order {order_name} is {mrca.name}.\")\n",
    "        mrca.name = order_name\n",
    "        print(f\"Renamed MRCA node to {mrca.name}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Finally, let's save the cleaned species tree to a file.\n",
    "\n",
    "# This is the first output - make sure the output directory exists and is\n",
    "# correct. This will be the cleaned tree with order names added.\n",
    "import os\n",
    "if not os.path.exists('output'):\n",
    "    os.makedirs('output')\n",
    "\n",
    "species_tree.write(format=1, outfile='output/Actinopterygii_species_with_order.nwk')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22",
   "metadata": {},
   "source": [
    "# Tree of orders\n",
    "\n",
    "We do need a tree that shows the relationships between the different orders. This is used in subsequent notebooks to build a scaffold tree that will have the points grafted onto it. See the insect dataset for more info on this.\n",
    "\n",
    "<b>IMPORTANT</b>\n",
    "\n",
    "You *can* generate an order-level tree directly from the timetree.org site, however it usually is paraphyletic; certain orders show up multiple times in the tree and it requires extensive cleaning (and knowledge of fish phylogenetics, which I do not have). So we will use the species-level tree from timetree to generate monophyletic order-level groupings and then generate an order-level tree from there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each order, pick one leaf at random and print the leaf name, order, and family.\n",
    "\n",
    "# Let's start with a full copy of the original tree that we can prune.\n",
    "fish_tree_order_level = species_tree.copy()\n",
    "\n",
    "fishes_to_keep = []\n",
    "\n",
    "import random\n",
    "for order in found_orders:\n",
    "    leaves_in_order = fish_tree_order_level.search_nodes(order=order)\n",
    "    if leaves_in_order:\n",
    "        random_leaf = random.choice(leaves_in_order)\n",
    "        fishes_to_keep.append(random_leaf)\n",
    "\n",
    "# Now prune the tree to keep only these leaves.\n",
    "fish_tree_order_level.prune(fishes_to_keep, preserve_branch_length=True)\n",
    "\n",
    "# Now change the tip names to just the order name.\n",
    "for leaf in fish_tree_order_level.get_leaves():\n",
    "    leaf.name = leaf.order\n",
    "\n",
    "# Now make sure any internal nodes that have the same name as a leaf are renamed to something else.\n",
    "for node in fish_tree_order_level.traverse():\n",
    "    if node.is_leaf():\n",
    "        continue\n",
    "    if node.name in found_orders:\n",
    "        node.name = f\"{node.name}_internal\"\n",
    "\n",
    "print(f'Order-level tree has {len(fish_tree_order_level.get_leaves())} tips.')\n",
    "print(fish_tree_order_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Looks good, let's save the order-level tree to a file.\n",
    "fish_tree_order_level.write(format=1, outfile=\"output/Actinopterygii_order_level.nwk\")\n",
    "# Let's load it back in to make sure it saved correctly.\n",
    "order_level_tree = Tree(\"output/Actinopterygii_order_level.nwk\", format=1, quoted_node_names=True)\n",
    "print(order_level_tree)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25",
   "metadata": {},
   "source": [
    "# Taxonomy lookup CSV\n",
    "\n",
    "Let's make a CSV file that has the following columns for use in later notebooks:\n",
    "\n",
    "- genus\n",
    "- order\n",
    "- family\n",
    "- taxon\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import pandas as pd\n",
    "genus_order_family_df = pd.DataFrame(columns=['genus', 'order', 'family', 'taxon'])\n",
    "\n",
    "for leaf in species_tree.get_leaves():\n",
    "    genus_name = re.split(r'[ _-]', leaf.name)[0]\n",
    "\n",
    "    order_name = leaf.order\n",
    "    family_name = db.get_species_info(genus_name, verbose=False)[1]\n",
    "    genus_order_family_df.loc[len(genus_order_family_df)] = [genus_name, order_name, family_name, leaf.name]\n",
    "\n",
    "# Save this list to a file.\n",
    "genus_order_family_df.to_csv('output/Actinopterygii_genus_order_family_taxon.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27",
   "metadata": {},
   "source": [
    "# INCOMPLETE - Tree of families.\n",
    "\n",
    "Let's do the same thing for families. We'll use this to build a scaffold, similar to the order-level tree, but using monophyletic families.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# First let's get a list of all the unique families in the tree.\n",
    "unique_families = set()\n",
    "for leaf in species_tree.get_leaves():\n",
    "    unique_families.add(leaf.family)\n",
    "print(f\"Found {len(unique_families)} unique families in the tree.\")\n",
    "for family in sorted(unique_families):\n",
    "    print(f\"- {family}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29",
   "metadata": {},
   "source": [
    "# (OPTIONAL) Family-level statistics\n",
    "\n",
    "The cells below do some histograms to get counts of taxa in families, etc. These are not needed for making OpenSpace assets, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's show a barplot of the number of leaves per order. Add the number of leaves on top of each bar.\n",
    "import matplotlib.pyplot as plt\n",
    "order_counts = {order: len([leaf for leaf in species_tree.get_leaves() if leaf.order == order]) for order in found_orders}\n",
    "# Put these in alphabetical order.\n",
    "#order_counts = dict(sorted(order_counts.items()))\n",
    "plt.figure(figsize=(10, 6))\n",
    "bars = plt.bar(order_counts.keys(), order_counts.values())\n",
    "plt.xticks(rotation=90)\n",
    "plt.xlabel('Order')\n",
    "plt.ylabel('Number of Leaves')\n",
    "\n",
    "# Add the number of leaves on top of each bar.\n",
    "for bar in bars:\n",
    "    yval = bar.get_height()\n",
    "    #plt.text(bar.get_x() + bar.get_width()/2, yval, int(yval), va='bottom')  # va: vertical alignment\n",
    "    plt.text(bar.get_x(), yval, int(yval), va='bottom')  # va: vertical alignment\n",
    "plt.title('Number of Leaves per Order in Species Tree')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# How many unique families in each order? Let's make a barplot.\n",
    "unique_families_per_order = genus_order_family_df.groupby('order')['family'].nunique()\n",
    "\n",
    "# Change the order of the list to match the found_orders list of fish orders\n",
    "unique_families_per_order = unique_families_per_order.reindex(found_orders)\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.figure(figsize=(10, 6))\n",
    "plt.bar(unique_families_per_order.index, unique_families_per_order.values)\n",
    "plt.xticks(rotation=90)\n",
    "plt.title('Number of Unique Families per Order')\n",
    "plt.xlabel('Order')\n",
    "plt.ylabel('Number of Unique Families')\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "insect_trees_orig",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
